accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1248,92 @@ the `LEGACY_ERASURE` of the types.
+are equal using the definition of runtime type equality specified above.  The
+object `R` which serves as the canonical form for `T1` and `T2` is defined as
+follows. Let `S1` and `S2` be the types to which `T1` and `T2` correspond.  Note
+that by assumption, **NORM**(`S1`) and **NORM**(`S2`) are syntactically
+identical up to the identity of bound variables and the placement of legacy type
+modifiers.  We choose `R` arbitrarily as the type literal corresponding to one
+of **NORM**(`S1`) or **NORM**(`S2`) with all legacy type modifiers erased.
+
+The effect of this definition is to ensure that type literals which look
+identical in the source syntax but which may differ by the presence of legacy
+type modifiers are canonicalized consistently in the sense that any two type
+literals which would compare equal via the definition of runtime type equality
+given above will be canonicalized to the same type object.  The choice of what
+type object to canonicalize to is arbitrary in that placement of legacy
+modifiers in type literals is not otherwise observable in the language.  Any
+consistent choice of placement of legacy modifiers is hence an equally valid
+implementation.
+
+Note that the choice of canonicalization for type literals does not depend
+directly on whether sound or unsound null safety semantics are in use.  However,
+soundly null safe programs will never contain legacy types, and hence the
+erasure of legacy type modifiers is never necessary when running in sound mode.
+
+#### Constant instances
+
+In both sound and unsound null checking, canonicalization of generic object
+instances is performed relative to the normal forms of their generic type
+arguments.  That is, an instance of `C<T0>` canonicalizes to the same object as
+`C<T1>` if `T0` and `T1` have the same normal form (up to the identity of bound
+variables), and the objects are otherwise identical.
 
 With sound null checking, all generic const constructors and generic const
-literals are evaluated using the actual type arguments provided, whether legacy
-or non-legacy.  This ensures that with sound null checking, the final
-consistent semantics are obeyed.
+literals are evaluated using the actual type arguments provided.  This ensures
+that with sound null checking, the final consistent semantics are obeyed.
+
+With unsound null checking, all generic const constructors and generic const
+literals are additionally treated as if all type arguments passed to them were
+legacy types regardless of whether the constructed class is defined in a legacy
+library or not, and regardless of whether the constructor invocation or literal
+occurs in a legacy library or not.  That is, a constructor invocation or object
+literal with generic type parameters `Ti` is treated as if the parameters were
+**LEGACY_TYPE**('Ti'). Implementations may choose to eagerly normalize the type
+arguments before applying the legacy rewrite, as desired.
```

Comment:
(I assume that type aliases are always expanded eagerly).

Reply:
Why do we not require eager normalization? It will occur at compile-time, so there's no run-time cost, and it seems to be a useless source of ambiguity.

If the outcome is provably the same as an approach where normalization is not performed eagerly, each implementation is of course free to use such an approach. This would allow the reader of the code to reason about canonicalization based on a simpler mental model than the one which is needed if normalization is optional.